mypyc/genops.py

```diff
@@ -1130,6 +1130,19 @@ def visit_index_expr(self, expr: IndexExpr) -> Value:
             return self.add(TupleGet(base, expr.index.value, expr.line))
 
         index_reg = self.accept(expr.index)
+        if isinstance(self.node_type(expr.base), RInstance):
+            typ = self.types[expr.base]
+            signature = self.get_native_method_signature(typ, '__getitem__')
+            if signature is not None:
+                args = [index_reg]
+                arg_types = [self.type_to_rtype(arg_type) for arg_type in signature.arg_types]
+                arg_regs = self.coerce_native_call_args(args, arg_types, expr.line)
+                return self.add(MethodCall(self.type_to_rtype(signature.ret_type),
+                                           base,
+                                           '__getitem__',
+                                           arg_regs,
+                                           expr.line))
+
```

Comment:
With this now having both an `RInstance` check and a `translate_special_method_call`, it ends up duplicating a lot of the logic of `translate_method_call`. Can you try factoring some code out of `translate_method_call` to produce a helper that can be used to easily invoke methods like this?